### DISCUSSION TITLE
Issue with two devices connected: "Event loop already running" with qasync, asyncio and PyQt5

### DISCUSSION THREAD

**Discussion by ivanjimenez** [hbldh/bleak#852] [Q&A]
* bleak version: 0.14.3
* Python version: 3.9
* Operating System: Windows 10 LTSBC
* BlueZ version (`bluetoothctl -v`) in case of Linux:
* Python for QT5

### Description

I'm trying to use this example in order to connect two inertials at the same time, but sometimes are connected and sometimes only is connected one of them. Then I click the connect button from my app and both are running. The behaviour is not always the same, sometimes two devices are connected well and if I stop the transfer and I want to execute another time the caputure, then fails for some inertial and I have to push the button.

I am using qasync and QEventloop in order to use asyncio for PyQt5.

The main file uses this code:

```
    try:
        loop = qasync.QEventLoop(app)
        asyncio.set_event_loop(loop)
    except asyncio.InvalidStateError as e:
        print(e)

    with loop:

        loop.run_forever()
```

And the code to run the capture:

```
 def run(self, addresses):

        loopIMU = asyncio.get_event_loop()
        loopIMU2 = asyncio.get_running_loop()

        for address in addresses:
            try:
                loopIMU.create_task(self.connect_to_device(address))

            except Exception as e:
                loopIMU.stop()
                address = address
                loopIMU2.create_task(self.connect_to_device(address))

        loopIMU.run_forever()
        loopIMU2.run_forever()

    async def connect_to_device(self, address):
        global gl_client
        async with BleakClient(address) as client:
            gl_client = client
            print(f"Connected: {address} {client.is_connected}")
            await asyncio.sleep(0.8)
            # Look closely at what the loop below says. It says Do `start_notify` forever. This is not what you want...
            # This will essentially stop and restart your notifications in very short time intervals forever
            # leading to unpredictable behaviour.
            while True:
                await client.start_notify(UART_TX_CHAR_UUID, partial(self.callback2, client))
                data = await client.read_gatt_char(UART_TX_CHAR_UUID)
                # print(data)
                await client.write_gatt_char(....)

and I use:

self.run(addresses) # where addreses = [mac1, mac2]

### What I Did


Always, when capturing is starting, it prints the following error:

Traceback (most recent call last):
  File "...", line 534, in runDevices
    self.run([address1, address2])  
  File "...", line 552, in run
    loopIMU.run_forever()
  File "...", line 369, in run_forever
    raise RuntimeError("Event loop already running")
RuntimeError: Event loop already running

I follow the code advice in this repository and issues. I think is a problem with qasync or asyncio. And I have checked various solutions without success.

I would appreciate some help with this issue.

Cheers.

**Comment by dlech** [comment]
As it says in the error message, you can only have one event loop running at a time. In general, the best way to do it is to use `asyncio.run()` at the entry point of the program.  For the `qasync` library, it looks like the top-level run would be `qasync.run()` instead of `asyncio.run()`.

**Reply by ivanjimenez** [reply]
Hi, Dlech:

You gave me an idea. I tested this method:

```
    def run(self, addresses):

        asyncio.ensure_future(self.connect_to_device(addresses[0]))
        asyncio.ensure_future(self.connect_to_device(addresses[1]))
```

And I haven't error with EventLoop. The problem is qasync is running a loop plus QEventloop of Qt and after two loops. Sure is better add tasks to actual event loop.

It's difficult use ayncio.run and qasync.run in QT, I didn't find a proper way to do that. Documentation for asyncio and qasync is terrible in my opinion.

So, I think with this is working. Remember: QT+Python+Qasync+asyncio. It's better add tasks to event loop. 

Anyway, if I have another update o issue I will follow this thread if you are agree.

Thanks.

**Reply by dlech** [reply]
Yes, this is the proper way to schedule tasks.

But connecting to two devices at the same time like this is known to be problematic. https://bleak.readthedocs.io/en/latest/troubleshooting.html#connecting-to-multiple-devices-at-the-same-time

**Reply by ivanjimenez** [reply]
Hi, David. My idea is using bleak for connecting 5 devices at the same time following two_Devices.py script and others from this repo. But I think if it is problematic for two devices, then bleak is not done for two or more devices. But it is possible that if you use a usb dongle for connect a number of devices, this problems won't be so huge. Or maybe I have to think in other library in this case. 

I appreciate your opinion.
Thanks a lot

**Reply by dlech** [reply]
Assuming the hardware supports it, it should be fine if you just scan and connect one device at time. The first device can stay connected when you scan and connect the second device and so on. The problems happen when `BleakClient` is used to do the scanning since it will try to start to instances of `BleakScanner` at the same time. This causes one scanner to still be running when the actually connection process is happening and the connection will fail sometimes. This can be avoided by making sure that there is no scanning during the actual connection process of one device. Then you can start scanning again.

**Reply by ivanjimenez** [reply]
Yes, I do that: only Scanning process previous connection. So it sounds great. I have been using two inertials at the same time but the problems were for USB 4.1. HW Manufacturer recommended me >= 5.0 for two or more devices at the same time. I think this module can be fine for my project. But, do you recommend me another open source libraries for Windows or maybe this is the best?

Kind Regards
Iván
